### ISSUE TITLE
Add A Learning Rate Finder Algorithm #reproduciblescience

### ISSUE THREAD

**Issue by maf88** [ivadomed/ivadomed#598]
### Motivation for the feature

In order to improve setting hyper-parameters, I propose that we add a learning rate finder algorithm. This way, we will always have an optimal initial LR according to loss functions, models, and  datasets.
 
### Description of the feature

This algorithm was proposed by: 

> "Smith LN. Cyclical Learning Rates for Training Neural Networks. arXiv [cs.CV]. 2015."

I implemented it for one of our repository, and it works very well in order to compare different training strategy and loss function to reconstruct data. This kind of algorithm ensure equal training condition for a deep-learning pipeline for different loss functions, models, or datasets. 
This algorithm should be used as a script, where you pass your dataloader, model, and loss function.
It has three main parameters: `LR_MIN`, `LR_MAX`, and `LR_ITERATIONS`.  
The `LR_MAX` and the `LR_MIN` must be modified until a step increase of the loss is 
noticed. The number of iterations is also an important parameter of this
algorithm since the model is trained during the LR scan. 
After some trial and error, a number of iterations between [200, 500], and a
minimal of 1/4 of the maximal LR yielded great results.
In Pytorch, you need to add a `scheduler.step()` after each mini-bach backpropagation. 

The biggest advantage of using a LR finder algorithm is to always reduce the number of epochs for training your ANN, while avoiding divergence problems.

### Alternatives

Other type of algorithms can be used, but they are very slow.

### Additional context

URL: [http://arxiv.org/abs/1506.01186](http://arxiv.org/abs/1506.01186)

**Comment by maf88** [comment]
For future reference: Pytorch has an argument auto_lr_find for the `Trainer` class that is also an implementation based on the article cited above. Definitely, it is worth it to see if it can be use as is with ivadomed.

The tutorial: [https://pytorch-lightning.readthedocs.io/en/latest/lr_finder.html](https://pytorch-lightning.readthedocs.io/en/latest/lr_finder.html)

**Comment by charleygros** [comment]
> In order to improve setting hyper-parameters, I propose that we add a learning rate finder algorithm. This way, we will always have an optimal initial LR according to loss functions, models, and datasets.

Love this! That would be a great addition @maf88!! All green lights :+1:
Please let us know if you would like some help with the integration of your script within `ivadomed`, happy to help!
